// appium-xcuitest-driver/build/lib/commands/permissions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

function requireSimulator(driver) {
  if (!driver.isSimulator()) {
    throw new Error('Permissions-related operations can only be performed on Simulator');
  }
}

function requireOptions(opts = {}) {
  if (!opts.bundleId) {
    throw new Error(`The 'bundleId' options must be a valid application bundle identifier`);
  }

  return opts;
}

commands.mobileGetPermission = async function mobileGetPermission(opts = {}) {
  const {
    service,
    bundleId
  } = requireOptions(opts);

  if (!service) {
    throw new Error(`The 'service' option is expected to be present`);
  }

  requireSimulator(this);
  return await this.opts.device.getPermission(bundleId, service);
};

commands.mobileSetPermissions = async function mobileSetPermissions(opts = {}) {
  const {
    access,
    bundleId
  } = requireOptions(opts);

  if (!_lodash.default.isPlainObject(access)) {
    throw new Error(`The 'access' option is expected to be a map`);
  }

  requireSimulator(this);
  await this.opts.device.setPermissions(bundleId, access);
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
