// nativescript-dev-appium/lib/element-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
const locators_1 = require("./locators");
class ElementHelper {
    constructor(_args) {
        this._args = _args;
        this.locators = new locators_1.Locator(this._args);
    }
    getXPathElement(name) {
        const tempName = name.toLowerCase().replace(/\-/g, "");
        return `//${this.locators.getElementByName(tempName)}`;
    }
    ;
    getXPathByText(text, exactMatch) {
        return this.findByTextLocator("*", text, exactMatch);
    }
    getXPathWithExactText(text) {
        return this.getXPathByText(text, true);
    }
    getXPathContainingText(text) {
        return this.getXPathByText(text, false);
    }
    findByTextLocator(controlType, value, exactMatch) {
        const result = this.getXPathByTextAttributes(`//${controlType}`, value, exactMatch);
        return result;
    }
    getXPathByTextAttributes(controlType, textValue, exactMatch) {
        let attributes = ["label", "value", "hint"];
        if (this._args.isAndroid) {
            attributes = ["content-desc", "resource-id", "text"];
        }
        let searchedString = "";
        if (exactMatch) {
            if (this._args.isAndroid) {
                attributes.forEach((atr) => { searchedString += "translate(@" + atr + ",'ABCDEFGHIJKLMNOPQRSTUVWXYZ','abcdefghijklmnopqrstuvwxyz')='" + textValue.toLowerCase() + "'" + " or "; });
            }
            else {
                attributes.forEach((atr) => { searchedString += "@" + atr + "='" + textValue + "'" + " or "; });
            }
        }
        else {
            if (this._args.isAndroid) {
                attributes.forEach((atr) => { searchedString += "contains(translate(@" + atr + ",'ABCDEFGHIJKLMNOPQRSTUVWXYZ','abcdefghijklmnopqrstuvwxyz'),'" + textValue.toLowerCase() + "')" + " or "; });
            }
            else {
                attributes.forEach((atr) => { searchedString += "contains(@" + atr + ",'" + textValue + "')" + " or "; });
            }
        }
        searchedString = searchedString.substring(0, searchedString.lastIndexOf(" or "));
        const result = `${controlType}[${searchedString}]`;
        utils_1.log("Xpath: " + result, this._args.verbose);
        return result;
    }
}
exports.ElementHelper = ElementHelper;
