// mobile-devices-controller/lib/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Platform;
(function (Platform) {
    Platform["ANDROID"] = "android";
    Platform["IOS"] = "ios";
})(Platform = exports.Platform || (exports.Platform = {}));
var DeviceType;
(function (DeviceType) {
    DeviceType["SIMULATOR"] = "simulator";
    DeviceType["EMULATOR"] = "emulator";
    DeviceType["DEVICE"] = "device";
    DeviceType["TV"] = "tv";
    DeviceType["WATCH"] = "watch";
})(DeviceType = exports.DeviceType || (exports.DeviceType = {}));
var Status;
(function (Status) {
    Status["SHUTDOWN"] = "shutdown";
    Status["BOOTED"] = "booted";
    Status["OFFLINE"] = "offline";
    Status["BUSY"] = "busy";
    Status["UNAUTORIZED"] = "unautorized";
    Status["INVALID"] = "invalid";
})(Status = exports.Status || (exports.Status = {}));
var AndroidKeyEvent;
(function (AndroidKeyEvent) {
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_UNKNOWN"] = 0] = "KEYCODE_UNKNOWN";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_MENU"] = 1] = "KEYCODE_MENU";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SOFT_RIGHT"] = 2] = "KEYCODE_SOFT_RIGHT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_HOME"] = 3] = "KEYCODE_HOME";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_BACK"] = 4] = "KEYCODE_BACK";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_CALL"] = 5] = "KEYCODE_CALL";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_ENDCALL"] = 6] = "KEYCODE_ENDCALL";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_0"] = 7] = "KEYCODE_0";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_1"] = 8] = "KEYCODE_1";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_2"] = 9] = "KEYCODE_2";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_3"] = 10] = "KEYCODE_3";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_4"] = 11] = "KEYCODE_4";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_5"] = 12] = "KEYCODE_5";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_6"] = 13] = "KEYCODE_6";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_7"] = 14] = "KEYCODE_7";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_8"] = 15] = "KEYCODE_8";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_9"] = 16] = "KEYCODE_9";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_STAR"] = 17] = "KEYCODE_STAR";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_POUND"] = 18] = "KEYCODE_POUND";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_DPAD_UP"] = 19] = "KEYCODE_DPAD_UP";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_DPAD_DOWN"] = 20] = "KEYCODE_DPAD_DOWN";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_DPAD_LEFT"] = 21] = "KEYCODE_DPAD_LEFT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_DPAD_RIGHT"] = 22] = "KEYCODE_DPAD_RIGHT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_DPAD_CENTER"] = 23] = "KEYCODE_DPAD_CENTER";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_VOLUME_UP"] = 24] = "KEYCODE_VOLUME_UP";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_VOLUME_DOWN"] = 25] = "KEYCODE_VOLUME_DOWN";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_POWER"] = 26] = "KEYCODE_POWER";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_CAMERA"] = 27] = "KEYCODE_CAMERA";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_CLEAR"] = 28] = "KEYCODE_CLEAR";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_A"] = 29] = "KEYCODE_A";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_B"] = 30] = "KEYCODE_B";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_C"] = 31] = "KEYCODE_C";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_D"] = 32] = "KEYCODE_D";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_E"] = 33] = "KEYCODE_E";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_F"] = 34] = "KEYCODE_F";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_G"] = 35] = "KEYCODE_G";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_H"] = 36] = "KEYCODE_H";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_I"] = 37] = "KEYCODE_I";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_J"] = 38] = "KEYCODE_J";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_K"] = 39] = "KEYCODE_K";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_L"] = 40] = "KEYCODE_L";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_M"] = 41] = "KEYCODE_M";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_N"] = 42] = "KEYCODE_N";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_O"] = 43] = "KEYCODE_O";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_P"] = 44] = "KEYCODE_P";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_Q"] = 45] = "KEYCODE_Q";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_R"] = 46] = "KEYCODE_R";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_S"] = 47] = "KEYCODE_S";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_T"] = 48] = "KEYCODE_T";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_U"] = 49] = "KEYCODE_U";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_V"] = 50] = "KEYCODE_V";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_W"] = 51] = "KEYCODE_W";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_X"] = 52] = "KEYCODE_X";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_Y"] = 53] = "KEYCODE_Y";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_Z"] = 54] = "KEYCODE_Z";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_COMMA"] = 55] = "KEYCODE_COMMA";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_PERIOD"] = 56] = "KEYCODE_PERIOD";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_ALT_LEFT"] = 57] = "KEYCODE_ALT_LEFT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_ALT_RIGHT"] = 58] = "KEYCODE_ALT_RIGHT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SHIFT_LEFT"] = 59] = "KEYCODE_SHIFT_LEFT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SHIFT_RIGHT"] = 60] = "KEYCODE_SHIFT_RIGHT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_TAB"] = 61] = "KEYCODE_TAB";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SPACE"] = 62] = "KEYCODE_SPACE";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SYM"] = 63] = "KEYCODE_SYM";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_EXPLORER"] = 64] = "KEYCODE_EXPLORER";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_ENVELOPE"] = 65] = "KEYCODE_ENVELOPE";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_ENTER"] = 66] = "KEYCODE_ENTER";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_DEL"] = 67] = "KEYCODE_DEL";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_GRAVE"] = 68] = "KEYCODE_GRAVE";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_MINUS"] = 69] = "KEYCODE_MINUS";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_EQUALS"] = 70] = "KEYCODE_EQUALS";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_LEFT_BRACKET"] = 71] = "KEYCODE_LEFT_BRACKET";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_RIGHT_BRACKET"] = 72] = "KEYCODE_RIGHT_BRACKET";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_BACKSLASH"] = 73] = "KEYCODE_BACKSLASH";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SEMICOLON"] = 74] = "KEYCODE_SEMICOLON";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_APOSTROPHE"] = 75] = "KEYCODE_APOSTROPHE";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SLASH"] = 76] = "KEYCODE_SLASH";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_AT"] = 77] = "KEYCODE_AT";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_NUM"] = 78] = "KEYCODE_NUM";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_HEADSETHOOK"] = 79] = "KEYCODE_HEADSETHOOK";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_FOCUS"] = 80] = "KEYCODE_FOCUS";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_PLUS"] = 81] = "KEYCODE_PLUS";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_MENU2"] = 82] = "KEYCODE_MENU2";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_NOTIFICATION"] = 83] = "KEYCODE_NOTIFICATION";
    AndroidKeyEvent[AndroidKeyEvent["KEYCODE_SEARCH"] = 84] = "KEYCODE_SEARCH";
    AndroidKeyEvent[AndroidKeyEvent["TAG_LAST_KEYCODE"] = 85] = "TAG_LAST_KEYCODE";
})(AndroidKeyEvent = exports.AndroidKeyEvent || (exports.AndroidKeyEvent = {}));
